---
title: "Exclude HW snps"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load in the Data

```{r cars}
library(snpStats)
library(pbmcapply)
library(VariantAnnotation)
```


```{r, echo=FALSE}
fl = list.files("/home/kgoldmann/Documents/PEAC_eqtl/Outputs/DNA/", full.names = TRUE)
fl = fl[grepl("_4PCA_all.vcf.gz", fl)]
fl = fl[! grepl("tbi", fl)]
```

## Find the snps which pass the Hardy Weinberg Value

# See here file:///home/katrionagoldmann/Downloads/Lab12%20(1).html
```{r}
HW_test = function(file){
  chr = gsub("/home/kgoldmann/Documents/PEAC_eqtl/Outputs/DNA//PEAC_|_4PCA_all.vcf.gz", "", file)
  print(chr)
  vcf = readVcf(file)
  res <- genotypeToSnpMatrix(vcf)
  gt = res$genotypes
  
  out = col.summary(gt) # Find the HW z values
  
  hardy <- 10^-6      # HWE cut-off
  HWEuse <- with(out, !is.na(z.HWE) & ( abs(z.HWE) < abs( qnorm(hardy/2) ) ) )
  HWEuse[is.na(HWEuse)] <- FALSE          # Remove NA's as well
  cat(ncol(gt)-sum(HWEuse),"SNPs will be removed due to high HWE.\n") 
  cat(sum(HWEuse),"SNPs will be kept.\n") 
  out2 = out[HWEuse, ]
  write.table(rownames(out2), file=paste0("/home/kgoldmann/Documents/PEAC_eqtl/Outputs/DNA/HW_keep_", chr, ".txt"), sep=",", quote=FALSE, col.names = FALSE, row.names = FALSE)
}
```


```{r}
set.seed(1)
fin = lapply(fl, function(x) HW_test(x))
````



